core/include/drivers/driver.h

```diff
@@ -0,0 +1,155 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/**
+ * @copyright 2018 NXP
+ *
+ * @file    driver.h
+ *
+ * @brief   Definition of the generic driver object opened, closed,
+ *          suspended and resumed by the system.
+ *
+ */
+#ifndef __DRIVER_H__
+#define __DRIVER_H__
+
+#include <scattered_array.h>
+#include <tee_api_types.h>
+#include <trace.h>
+
+#if TRACE_LEVEL >= TRACE_DEBUG
+#define DRV_MGT_DEBUG
+#endif
+
+#ifndef CFG_PLAT_PWR
+/**
+ * @brief   Enumerate the driver power mode state
+ */
+enum drv_pwrmode {
+	ON = 0,   ///< Power up and running
+	IDLE,     ///< Low Power mode Idle
+	OFF       ///< Power OFF
+};
+#else
+#include <platform_pwr.h>
+#endif
+
+/**
+ * @brief   Generic driver object operations. Some can be optional.
```

Comment:
@cneveux take your pick! Any will do I suppose, as long as it illustrates why you are implementing things like you do. If you have several use cases with more complex power states involved (you mentioned clock gating for instance), then it's even better :)

Reply:
For info, in the upstream plans for the stm32mp1 platfrom, there is a pm framework. It is not planned to be generic but maybe you could do something with, even if I think it does not suit your needs. Here is the code.

PM framework header file:
```C
/*
 * Drivers can register a callback for the suspend and resume sequences and
 * a private cookie passed as argument to the callback.
 * The same callback is used for suspend and resume. First argument of the
 * callback defines whether the device shall suspend or resume.
 *
 * Driver should tag its callback resource as unpaged (i.e KEEP_PAGER())
 * since the callback is called from an unpaged execution context.
 */
enum pm_op {
	PM_OP_SUSPEND,
	PM_OP_RESUME,
};

void stm32mp_register_pm_cb(void (*callback)(enum pm_op op, void *handle),
			    void *handle);
```
PM framework implementation:
```C
static bool __maybe_unused pm_cb_is_valid(void (*cb)(enum pm_op op, void *hdl),
					  void *hdl)
{
	paddr_t cb_phy = virt_to_phys(cb);
	paddr_t hdl_phy = virt_to_phys(hdl);
	bool __maybe_unused valid;

	valid = (phys_to_virt(cb_phy, MEM_AREA_TEE_RAM_RX) == cb) &&
		((phys_to_virt(hdl_phy, MEM_AREA_TEE_RAM_RX) == hdl) ||
		 (phys_to_virt(hdl_phy, MEM_AREA_TEE_RAM_RO) == hdl) ||
		 (phys_to_virt(hdl_phy, MEM_AREA_TEE_RAM_RW) == hdl));

	if (!valid) {
		EMSG("pm_cb mandates unpaged arguments %p %p", (void *)cb, hdl);
	}

	return valid;
}

struct pm_cb {
	void (*callback)(enum pm_op op, void *handle);
	void *handle;
};
static struct pm_cb *pm_cb;
static size_t pm_cb_count;

void stm32mp_register_pm_cb(void (*callback)(enum pm_op op, void *handle),
			    void *handle)
{
#ifdef CFG_WITH_PAGER
	assert(pm_cb_is_valid(callback, handle));
#endif 

	pm_cb_count++;
	pm_cb = realloc(pm_cb, sizeof(struct pm_cb) * pm_cb_count);
	if (!pm_cb){
		panic();
	}

	pm_cb[pm_cb_count - 1].callback = callback;
	pm_cb[pm_cb_count - 1].handle = handle;
}

// Call this from the suspend sequence
static void save_soc_context(void)
{
	size_t n;

	for (n = 0; n < pm_cb_count; n++) {
		DMSG("call %p(%p)", (void *)pm_cb[n].callback, pm_cb[n].handle);
		pm_cb[n].callback(PM_OP_SUSPEND, pm_cb[n].handle);
	}
}

// Call this from the resume sequence
static void restore_soc_context(void)
{
	size_t n;

	/* Resume secure clocks before any driver is resumed */
	stm32mp1_clock_resume();

	for (n = 0; n < pm_cb_count; n++) {
		DMSG("call %p(%p)", (void *)pm_cb[n].callback, pm_cb[n].handle);
		pm_cb[n].callback(PM_OP_RESUME, pm_cb[n].handle);
	}
}
```
Driver FOO registers for pm:
```C
static void foo_suspend_resume(enum pm_op op, void *handle)
{
	switch(op) {
	case PM_OP_SUSPEND:
		//do something
		break;
	case PM_OP_RESUME:
		//do something
		break;
	}
}

// from some init-routine
	(...)
	stm32mp_register_pm_cb(foo_suspend_resume, NULL);
	(...)
```